Document/0x07a-Testing-Authentication-and-Session-Management.md

```diff
@@ -66,6 +66,106 @@ If any of these two conditions raise an issue, reject the request and do not all
+* Usage of asymmetric algorithms:
+  * Within JWT asymmetric algorithms as RSA or ECDSA can be used for signing the token. Due to vulnerabilities in some JWT frameworks the public key can be misused as HMAC secret key in order to sign the token.
+* Token Storage on client side:
+  * When using a mobile app that uses JWT it should be verified where the token is stored locally on the device<sup>[5]</sup>.
+* Cracking the signing key:
+  * Creating a signature of the token is done through a secret key on server side. Once a JWT is obtained there are several tools available that can try to brute force the secret key offline<sup>[8]</sup>. See the tools section for details.
+* Information Disclosure:
+  * Decode the Base-64 encoded JWT and check what kind of data is transmitted within it.
+
+Please also follow the test cases in the OWASP JWT Cheat Sheet<sup>[4]</sup> if JWT is used.
+
+#### Remediation
+
+The latest version available of the JWT libraries in use should be implemented, to avoid known vulnerabilities.
+
+Store the JWT on the mobile phone using a secure mechanism, like KeyChain on iOS or KeyStore on Android.
+
+If replay attacks are a risk for the app, `jti` (JWT ID) claim should be implemented.
+
+Ideally the content of JWT should be encrypted in order to ensure the confidentially of the information contained within it. There might be description of roles, usernames or other sensitive information available that should be protected. An example implementation in Java can be found in the OWASP JWT Cheat Sheet<sup>[4]</sup>
+
+#### References
+
+##### OWASP Mobile Top 10 2016
+
+* M4 - Insecure Authentication - https://www.owasp.org/index.php/Mobile_Top_10_2016-M4-Insecure_Authentication
+
+##### OWASP MASVS
+
+* 4.1: "If the app provides users with access to a remote service, an acceptable form of authentication such as username/password authentication is performed at the remote endpoint."
+
+##### CWE
+
+* CWE-287: Improper Authentication - https://cwe.mitre.org/data/definitions/287.html
+
+##### Info
+
+* [1] RFC 7519 JSON Web Token (JWT) - https://tools.ietf.org/html/rfc7519
+* [2] PHP JWT - https://github.com/firebase/php-jwt
+* [3] Java Spring with JWT - http://projects.spring.io/spring-security-oauth/docs/oauth2.html
+* [4] OWASP JWT Cheat Sheet - `https://www.owasp.org/index.php/JSON_Web_Token_(JWT)_Cheat_Sheet_for_Java`
+* [5] Sample of JWT Token - https://jwt.io/#debugger
+* [6] Critical Vulnerabilities in JSON Web Token - https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries/
+* [7] JWT the right way - https://stormpath.com/blog/jwt-the-right-way
+* [8] Attacking JWT Authentication - https://www.sjoerdlangkemper.nl/2016/09/28/attacking-jwt-authentication/
+
+##### Tools
+* jwtbrute - https://github.com/jmaxxz/jwtbrute
+* crackjwt - https://github.com/Sjord/jwtcrack/blob/master/crackjwt.py
+* John the ripper - https://github.com/magnumripper/JohnTheRipper
+
```

Comment:
> (if logout has been allowed) should make the token invalid 

This will be covered in the test case "### Testing the Logout Functionality". Will make a reference to it.

> copying a token to another device should / or should not make an attacker able to continue authenticated

Will make a reference to it.

> nice-to-have: maybe check if we can relate it to other specs that are using jwt tokens? because this is mostly about how the token is setup, but not about how a token can be used in a context of oauth and stuff like that :).

@cpholguera is working on a general description at the beginning of the chapter.